proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,414 @@
+      listed above. See `below <#constraints-special-syntax>`_.)
+
+   #. A use of ``(# ... | ... | ... #)`` is now disallowed.
+
+   #. An occurrence of ``GHC.Tuple.Tuplen ty1 ty2 ... tyn`` is pretty-printed as ``Tuple [ty1, ty2, ..., tyn]``.
+
+   #. An occurrence of ``GHC.Tuple.CTuplen ty1 ty2 ... tyn`` is pretty-printed as ``Constraints [ty1, ty2, ..., tyn]``.
+
+#. Three releases after this proposal is implemented, remove the ``Solo`` pattern synonym from ``GHC.Tuple``.
+
+Effect and Interactions
+-----------------------
+1. With ``-XListTuplePuns`` (which is on by default), all programs that are accepted today continue
+   to be accepted, and with the same meanings. Note that the peculiar dance around type tuples and constraint
+   tuples exists today; I have tried to describe the current implementation faithfully, above.
+
+#. With ``-XListTuplePuns`` (which is on by default), most pretty-printing will happen as it does
+   today. The exception is around unsaturated ``CTuplen``, which is not handled above. It is hard to have
+   an unsaturated constraint tuple, but possible by the use of a type family that decomposes one. Today's
+   GHC prints out e.g. ``ghc-prim-0.6.1:GHC.Classes.(%,%)``. Switching to ``GHC.Classes.CTuple2`` (which is
+   actually parseable) seems a positive improvement.
+
+#. With the definitions above, users can avoid puns in their lists and tuples.
+
+   .. _constraints-special-syntax:
+
+#. Note that the type syntax ``(ty1, ty2, ..., tyn) => ...`` is already special syntax. The parser does *not*
+   parse a type to the left of the ``=>``. This syntax thus remains completely unaffected by ``-XListTuplePuns``
+   and will continue to work with ``-XNoListTuplePuns``. Furthermore, because a type like ``(ty1, ty2, ... tyn) => ...``
+   does not contain any uses of ``CTuplen``, it will also continue to pretty-print just as today.
+
+   On the other hand, collections of constraints occurring not to the left of a ``=>`` are affected by
+   this proposal, for example in ``Dict (Eq a, Show b)`` (which would be written ``Dict (Constraints [Eq a, Show b])``
+   under this proposal). Another example is ``(Eq a, (Show a, Read a)) => a -> a``, which would not
+   be accepted under ``-XNoListTuplePuns``. Instead, the user should flatten the constraints or
+   write ``(Eq a, Constraints [Show a, Read a]) => a -> a``.
+
+#. An instance declaration like ``instance (C a, C b) => C (Tuple [a, b]) where ...`` would be
+   rejected because it uses a type family in the instance head. We might choose to relax
+   this restriction, by allowing type families in an instance head, as long as they can
+   reduce to a ground (i.e. type-family-free) type. This proposal does *not* include such
+   a lifting of the restriction, as the workaround is straightforward: just write
+   ``instance (C a, C b) => C (Tuple2 a b) where ...``. Still, we may decide to revisit
+   this in the future.
```

Comment:
Alternatively, you don't define `Tuple2` and such (at least not primitively), and make `Tuple ts` a data family instead. Would this work?

Reply:
No, because you can't translate what is now `instance Functor ((,) a)`.